asm/osMapTLBRdb.s

```diff
@@ -9,26 +10,27 @@
 
 .balign 16
 
-glabel osMapTLBRdb
-/* 0086E0 80007AE0 40085000 */  mfc0  $t0, $10
-/* 0086E4 80007AE4 2409001F */  li    $t1, 31
-/* 0086E8 80007AE8 40890000 */  mtc0  $t1, $0
-/* 0086EC 80007AEC 40802800 */  mtc0  $zero, $5
-/* 0086F0 80007AF0 240A0017 */  li    $t2, 23
-/* 0086F4 80007AF4 3C09C000 */  lui   $t1, 0xc000
-/* 0086F8 80007AF8 40895000 */  mtc0  $t1, $10
-/* 0086FC 80007AFC 3C098000 */  lui   $t1, 0x8000
-/* 008700 80007B00 00095982 */  srl   $t3, $t1, 6
-/* 008704 80007B04 016A5825 */  or    $t3, $t3, $t2
-/* 008708 80007B08 408B1000 */  mtc0  $t3, $2
-/* 00870C 80007B0C 24090001 */  li    $t1, 1
-/* 008710 80007B10 40891800 */  mtc0  $t1, $3
-/* 008714 80007B14 00000000 */  nop   
-/* 008718 80007B18 42000002 */  tlbwi 
-/* 00871C 80007B1C 00000000 */  nop   
-/* 008720 80007B20 00000000 */  nop   
-/* 008724 80007B24 00000000 */  nop   
-/* 008728 80007B28 00000000 */  nop   
-/* 00872C 80007B2C 40885000 */  mtc0  $t0, $10
-/* 008730 80007B30 03E00008 */  jr    $ra
-/* 008734 80007B34 00000000 */   nop   
+LEAF(osMapTLBRdb)
+    mfc0    $t0, C0_ENTRYHI
+    li      $t1, NTLBENTRIES
+    mtc0    $t1, C0_INX
+    mtc0    $zero, C0_PAGEMASK
+    li      $t2, (TLBLO_UNCACHED | TLBLO_D | TLBLO_V | TLBLO_G)
+    li      $t1, K2BASE
+    mtc0    $t1, C0_ENTRYHI
+    li      $t1, K0BASE
+    srl     $t3, $t1, TLBLO_PFNSHIFT
+    or      $t3, $t3, $t2
+    mtc0    $t3, C0_ENTRYLO0
+    li      $t1, TLBLO_G
+    mtc0    $t1, C0_ENTRYLO1
+    nop
+    tlbwi
+    nop
+    nop
+    nop
+    nop
+    mtc0    $t0, C0_ENTRYHI
```

Comment:
You were indeed looking at the right thing. For `tlbwi` and `tlbwr` the destination hazard is "TLB". That is, any instruction using the TLB will be subject to hazards following either of these two instructions (i.e. all load/store instructions use the TLB, see page 101 of the manual for a dissection of the pipeline operation of `lw`). The destination hazard is 8 PClock cycles, so using the calculation in the manual you would need 7 instructions in the worst case, and the worst case should be taken as we do not know what awaits after the function returns. The 4 extra nops are to make it so that the first instruction after the return is the 8th instruction, so there is no hazard. Similar should apply to unmap, although I believe there is one extra nop there that this doesn't explain.

Reply:
To get 7,
you get the 8 from the TLB Destination Hazard associated to tlbwi
and subtract (0 + 1)
and I'm not sure where does the 0 come from, the lowest source TLB-related is "Instruction fetch" and it's only at 2? what is the worse case than this